core/src/connection/manager/task.rs

```diff
@@ -194,149 +198,157 @@ where
                                 return Poll::Pending
                             }
-                            Poll::Ready(Ok(connection::Event::Handler(event))) => {
-                                this.state = State::EstablishedReady {
-                                    connection: Some(connection),
-                                    event: Event::Notify { id, event }
+                            Poll::Ready(Ok(Some(connection::Event::Handler(event)))) => {
+                                this.state = State::Established {
+                                    connection,
+                                    event: Some(Event::Notify { id, event }),
                                 };
-                                continue 'poll
+                                continue
                             }
-                            Poll::Ready(Ok(connection::Event::AddressChange(new_address))) => {
-                                this.state = State::EstablishedReady {
-                                    connection: Some(connection),
-                                    event: Event::AddressChange { id, new_address }
+                            Poll::Ready(Ok(Some(connection::Event::AddressChange(new_address)))) => {
+                                this.state = State::Established {
+                                    connection,
+                                    event: Some(Event::AddressChange { id, new_address }),
                                 };
-                                continue 'poll
+                                continue
+                            }
+                            Poll::Ready(Ok(None)) => {
+                                // The connection is closed, don't accept any further commands
+                                // and terminate the task with a final event.
+                                this.commands.get_mut().close();
+                                let event = Event::Closed { id: this.id, error: None };
+                                this.state = State::Terminating(event);
+                                continue
                             }
                             Poll::Ready(Err(error)) => {
-                                // Notify the manager of the error via an event,
-                                // dropping the connection.
-                                let event = Event::Error { id, error };
-                                this.state = State::EstablishedReady { connection: None, event };
-                                continue 'poll
+                                // Don't accept any further commands.
+                                this.commands.get_mut().close();
+                                // Terminate the task with the error, dropping the connection.
+                                let event = Event::Closed { id, error: Some(error) };
+                                this.state = State::Terminating(event);
+                                continue
```

Comment:
Why those `continue` statements in this `else` branch?

Reply:
I added them just for clarity, i.e. to one does not need to check if there is code following the enclosing `if` or `match` to know where execution continues.

Reply:
Seeing a `continue` makes me think there is code after the block that should be skipped and to my surprise there was none.

Reply:
That's an understandable perspective :upside_down_face: Since its not clearly helping then, I removed them again in https://github.com/libp2p/rust-libp2p/pull/1682/commits/6e91e543339c8ecd8c84e3becc50db206105641d.